gateway/src/p2p.rs

```diff
@@ -0,0 +1,237 @@
+    meta_exchange::{
+        self, Behaviour as MetaExchangeBehaviour, Event as MEEvent, Response as MEResponse, Role,
+    },
+};
+use futures::StreamExt;
+use libp2p::dns::Transport as DnsTransport;
+use libp2p::{
+    Multiaddr, PeerId,
+    core::{transport::Transport as _, upgrade},
+    identity,
+    noise::{Config as NoiseConfig, Error as NoiseError},
+    ping::{Behaviour as PingBehaviour, Config as PingConfig, Event as PingEvent},
+    swarm::{Config as SwarmConfig, NetworkBehaviour, Swarm, SwarmEvent},
+    tcp::{Config as TcpConfig, tokio::Transport as TcpTokioTransport},
+    yamux::Config as YamuxConfig,
+};
+use libp2p_request_response::{Message as RequestResponseMessage, ProtocolSupport};
+use log::{debug, error, info, warn};
+use std::{collections::HashSet, time::Duration};
+use tokio::sync::mpsc::{self, UnboundedReceiver, UnboundedSender};
+
+#[derive(NetworkBehaviour)]
+#[behaviour(out_event = "GatewayEvent")]
+struct GatewayBehaviour {
+    ping: PingBehaviour,
+    request_response: GMBehaviour,
+    meta_exchange: MetaExchangeBehaviour,
+}
+pub enum GatewayEvent {
+    Ping(PingEvent),
+    RequestResponse(GMEvent),
+    MetaExchange(MEEvent),
+}
+
+impl From<PingEvent> for GatewayEvent {
+    fn from(e: PingEvent) -> Self {
+        GatewayEvent::Ping(e)
+    }
+}
+
+impl From<GMEvent> for GatewayEvent {
+    fn from(e: GMEvent) -> Self {
+        GatewayEvent::RequestResponse(e)
+    }
+}
+
+impl From<MEEvent> for GatewayEvent {
+    fn from(e: MEEvent) -> Self {
+        GatewayEvent::MetaExchange(e)
+    }
+}
+
+/// Internal structure that holds channels for inter-module communication
+struct Channels {
+    tx_request: UnboundedSender<Request>,
+    rx_request: UnboundedReceiver<Request>,
+    tx_response: UnboundedSender<Response>,
+    rx_response: Value<UnboundedReceiver<Response>>,
+}
+
+pub struct P2PChannels {
+    pub tx_request: UnboundedSender<Request>,
+    pub rx_response: UnboundedReceiver<Response>,
+}
+
+enum Value<T> {
+    Here(T),
+    Moved,
+}
+
+pub struct P2P {
+    pub peer_id: PeerId,
+
+    node_urls: Vec<String>,
+
+    swarm: Swarm<GatewayBehaviour>,
+    channels: Channels,
+}
+
+impl P2P {
+    pub fn new(node_urls: Vec<String>) -> Result<P2P, Box<dyn std::error::Error>> {
+        let kp = identity::Keypair::generate_ed25519();
+        let peer_id = PeerId::from(kp.public());
+        info!("Local peer id: {:?}", peer_id);
+
+        let auth_config =
+            NoiseConfig::new(&kp).map_err(|e: NoiseError| format!("noise config error: {}", e))?;
+
+        let transport = TcpTokioTransport::new(TcpConfig::default().nodelay(true))
+            .upgrade(upgrade::Version::V1)
+            .authenticate(auth_config)
+            .multiplex(YamuxConfig::default())
+            .boxed();
+
+        let behaviour = GatewayBehaviour {
+            ping: PingBehaviour::new(
+                PingConfig::new()
```

Comment:
Heh, I wanted to comment that the dot-notation `.with_*()` syntax is more like `::build()`, not `::new()`, only to realize it's library code, not yours =)